src/main/java/com/Acrobot/ChestShop/Listeners/Item/ItemMoveListener.java

```diff
@@ -16,14 +25,25 @@ public class ItemMoveListener implements Listener {
 
     @EventHandler(ignoreCancelled = true, priority = EventPriority.HIGHEST)
     public static void onItemMove(InventoryMoveItemEvent event) {
-        if (event.getSource() == null || getHolder(event.getDestination(), false) instanceof BlockState) {
-            return;
-        }
+        InventoryHolder destinationHolder = getHolder(event.getDestination(), false);
+        InventoryHolder sourceHolder = getHolder(event.getSource(), false);
 
-        if (!ChestShopSign.isShopBlock(getHolder(event.getSource(), false))) {
-            return;
-        }
+        if (!(destinationHolder instanceof BlockState) && ChestShopSign.isShopBlock(sourceHolder)) {
+            event.setCancelled(true);
+        } else if (ChestShopSign.isShopBlock(destinationHolder) && sourceHolder instanceof Hopper) {
+            Block shopBlock = ChestShopSign.getShopBlock(destinationHolder);
+            Sign connectedSign = uBlock.getConnectedSign(shopBlock);
+
+            Inventory tempInv = Bukkit.createInventory(null, destinationHolder.getInventory().getSize() + 9);
+            tempInv.setContents(ItemUtil.deepClone(destinationHolder.getInventory().getContents()));
+            tempInv.addItem(event.getItem().clone());
+
+            StockCounterModule.updateCounterOnQuantityLine(connectedSign, tempInv);
 
-        event.setCancelled(true);
+            tempInv.clear();
+            tempInv.close();
```

Comment:
What's the reason for deep cloning the inventory? As far as I can tell the StockCounterModule doesn't do that itself either?

Also in general I would kinda prefer this logic to be in the StockCounterModule so that it's all in one place (maybe not a separate listener to avoid getting the InventoryHolder twice but maybe you could add a method like `updateOnItemMoveEvent` which takes the holders and does all the actual checking and update logic in the `StockCounterModule`? (of course only if it's enabled see comment above)

Reply:
What could modify the inventory? Did you actually see that happening? Because as far as I'm aware the methods only get the sizes, not modify anything. (Otherwise the already existing stock counters would change the inventory too)

And yes, please move it into the stock counter module. That also makes it clearer what will happen with the inventory (and make it easier to spot that nothing is meant to modify it too)

Reply:
What could modify the inventory? Well when you take the contents of an inventory as `ItemStack[] getContents()` and then use `setContents(ItemStack[])` on the other inventory it doesnt like that the same stacks are in two inventories at once so it sometimes takes them out of one of them. This has been a problem since long ago and I don't know if they ever fixed it so I just deep clone it to be safe.

I've moved the logic to `StockCounterModule#updateCounterOnItemMoveEvent`

If there is anything else I can help you with please let me know!

Reply:
Well I have to modify the inventory, since I have to add the items that would be added by the event.
Tho I understand this an expensive action, so I'll find a way to work without the deep clone.